mars/core/graph/entity.py

```diff
@@ -74,6 +84,20 @@ def results(self):
     def results(self, new_results):
         self._result_tileables = new_results
 
+    @property
+    def logic_key(self):
+        if not hasattr(self, "_logic_key") or self._logic_key is None:
+            token_keys = []
+            for node in self.topological_iter():
```

Comment:
`topological_iter` does not guarantee order for nodes.

Reply:
Year, so sort the `token_keys` in line 93.

Reply:
Then `topological_iter` can even be replaced with a BFS to save time.

Reply:
Good suggestion. Thanks a lot.